### PR TITLE
[CSM-419] Count fees in ADA, not lovelaces

### PR DESCRIPTION
**Pull request by flyingleafe**
Things which are done:

- [x] Fees formula is fixed
- [x] Payload generation is refactored and tests are passing now
- [x] Now there's enough money to test transaction sending in dev mode with big fees enabled

### CODE REVIEW

**Comment by gromakovsky** [comment]
Why this way and not vice-versa? It would be much easier to change constants. Currently core code doesn't know anything about ADA and everybody is happy. Why do we need extra complication?

**Comment by flyingleafe** [comment]
@gromakovsky it doesn't really matter where do we multiply constants by a million - I left constants as they are to avoid any confusion when e.g. Lars or Duncan come back to review and adjust these constants.
Though if you really don't like it this way, it's really easy to change it; most of the work here is related to payload generation anyway.

**Comment by gromakovsky** [comment]
> it doesn't really matter where do we multiply constants by a million - I left constants as they are to avoid any confusion when e.g. Lars or Duncan come back to review and adjust these constants.

I think the easiest and best way to avoid confusion, is to write comments in `constants.yaml` file explicitly stating that the values are in lovelaces, not in ADA.
If they are in ADA, it's possible to confuse them with lovelaces (especially for someone who was working only on core, where we only have `Coin`).

AFAIK, the only place where it's reflected in code is `calculateTxSizeLinear`, so I don't really mind if there is only one such place (though I still see some inconsistency).

Initially the description contained only one item (`Fees formula is fixed`) and when I saw 8 changed files I was afraid that you had to change 8+ places just to update this formula.

tl;dr: I don't mind having 1e6 multiple in only one place (though I see some inconsistency), but I think that there must be comments in `constants.yaml` file.

**Comment by gromakovsky** [comment]
Another alternative is to put this logic into aeson instances.

File: node/src/Pos/Genesis.hs [hunk 1]
```diff
@@ -310,9 +310,9 @@ devGenesisContext distr =
 
     -- HD wallets
     hdwSize = 2 -- should be positive
-    -- 200 coins split among hdwSize users. Should be small sum enough
-    -- to avoid making wallets slot leaders.
-    hdwDistr = FlatStakes (fromIntegral hdwSize) (mkCoin 200)
+    -- 20 ADA (20 millon coins) split among hdwSize users.
+    -- Shouldn't mess with LRC after CSL-1502
```
Comment (gromakovsky) [inline node/src/Pos/Genesis.hs]: DId you test it in demo?

Comment (flyingleafe) [reply node/src/Pos/Genesis.hs]: Yes, of course, works fine.

File: core/constants.yaml [hunk 2]
```diff
@@ -28,9 +28,11 @@ dev: &dev
   genesisSoftforkMin: 0.6 # 60% of total stake
   genesisSoftforkDec: 0.05 # 5% of total stake
   genesisTxFeePolicy:
+    # Note: these constants here are provided in lovelaces, not ADA
+    # (1 ADA is 1,000,000 lovelaces)
     txSizeLinear:
-      a: 0.155381 # absolute minimal fees per transaction
-      b: 4.3946e-5 # additional minimal fees per byte of transaction size
+      a: 155381 # absolute minimal fees per transaction
+      b: 43.946 # additional minimal fees per byte of transaction size
```
Comment (flyingleafe) [inline core/constants.yaml]: I decided to finally change constants instead of changing formula

File: node/src/Pos/Generator/Block/Payload.hs [hunk 3]
```diff
@@ -162,95 +162,55 @@ genTxPayload = do
...
+        let adder hm TxOutAux { toaOut = TxOut {..} } =
+                HM.insertWith (+) txOutAddress (coinToInteger txOutValue) hm
+            utxoBalances = foldl' adder mempty utxo
+            hasMoney addr = fromMaybe 0 (HM.lookup addr utxoBalances) > 0
+            addrsWithMoney = filter hasMoney utxoAddresses
+
+        -- Select input and output addresses
+
+        inputAddrs <- selectSomeFromList (1, 3) addrsWithMoney
+        outputAddrs <- selectSomeFromList (1, 3) utxoAddresses
```
Comment (volhovM) [inline node/src/Pos/Generator/Block/Payload.hs]: Why `(1,3)` ? 😕 
Shouldn't you be taking `tgpMaxOutputs` into account?

File: node/test/Test/Pos/Block/Logic/Mode.hs [hunk 4]
```diff
@@ -151,10 +151,12 @@ genSuitableStakeDistribution :: Word -> Gen StakeDistribution
 genSuitableStakeDistribution stakeholdersNum =
     oneof [ genFlat
           {-, genBitcoin-} -- is broken
-          , pure $ safeExpStakes (15::Integer) -- 15 participants should be enough
+          , pure $ safeExpStakes (25::Integer) -- 25 participants should be enough
```
[the remark this answers is not in the document]
Comment (flyingleafe) [reply node/test/Test/Pos/Block/Logic/Mode.hs]: Function `safeExpStakes` accepts only one argument, which is both number of stakeholders _and_ minimal exponent included in distribution. If the minimal exponent is 15, then the poorest stakeholder has only 2^15 = 32768 lovelaces, which is less than minimal transaction fee; thus, this stakeholder cannot send a single transaction. Increasing the minimal exponent ten-fold makes the case where we have not enough money to generate a transaction with valid fee almost improbable.

**Comment by dcoutts** [comment]
Looks ok to me, but I'm not properly qualified to review this bit of code.
